janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/BufferUtil.java

```diff
@@ -57,6 +65,51 @@ public static StaticBuffer getIntBuffer(int[] ids) {
         return StaticArrayBuffer.of(arr);
     }
 
+    /**
+     * We always add a byte marker at the beginning to indicate it's a string buffer. Then we
+     * write the string uncompressed. Finally, if the total buffer size is 8, we add a dummy byte
+     * at the end so that upon read time, JanusGraph knows the buffer does not store a long value.
+     * See {@link IDManager#getKeyID(StaticBuffer)} for more details.
+     *
+     * @param s
+     * @return
+     */
+    public static StaticBuffer getStringBuffer(String s) {
+        if (StringUtils.isEmpty(s) || !StringEncoding.isAsciiString(s)) {
+            throw new IllegalArgumentException("value must be non-empty ASCII string but received: " + s);
+        }
+        ByteBuffer buffer;
+        if (s.length() + byteSize == longSize) {
+            buffer = ByteBuffer.allocate(longSize + byteSize);
+        } else {
+            buffer = ByteBuffer.allocate(s.length() + byteSize);
+        }
+        buffer.put(STRING_ID_MARKER);
+        for (int i = 0; i < s.length(); i++) {
+            int c = s.charAt(i);
+            assert c <= 127;
+            byte b = (byte)c;
+            /**
+             * we don't have to apply STOP_MASK here because a static buffer
+             * has a fixed length, and thus upon read time, we know where
+             * to stop reading the string. This is more to keep it consistent
+             * with {@link StringEncoding#writeAsciiString(byte[], int, String)}
+             * where we use STOP_MASK to mark the end of the string.
+             *
+             * An additional benefit of doing so is to enable corruption check
+             * in the future. Note: this approach has no overhead.
+             */
+            if (i+1==s.length()) b |= STOP_MASK;
+            buffer.put(b);
+        }
```

Comment:
It feels that we could use UTF-8 characters instead of ASCII characters if we get rid of `STOP_MASK` here.
We lose 1 bit per character here anyway, so why not let users decide if they want to use ASCII characters or some other unicodes using UTF-8 encoding?
In such case, if we use UTF-8, we lose the ability to implement data corruption check in the future, but it seems to be more related to the storage backend features rather then to JanusGraph layer anyway (not sure here actually).
Moreover, this will simplify this method copy part to something like:
`System.arraycopy(s.getBytes(), 0, arr, 0, s.length());`

Moreover, UTF-8 might potentially be better here because if we want to use non-ASCII characters latter - we won't be able due to `String` type being reserved for ASCII character only. With UTF-8 it's up to the user what characters to use.

Notice, I personally think ASCII characters along are usually enough for IDs (I haven't seen people use other characters for IDs previously), but I just try to view on this problem from different perspective.

Reply:
Good point! My primary design principle is to reduce storage overhead as much as possible. An ordinary Java string uses 2 bytes for 1 character. UTF-8, even worse, uses 1 to 4 bytes per character.

If one would like to use any character other than visible ASCII characters, they could use base64 encoding on the app side, or they could just use a byte array (once we support it). Since ASCII characters are usually enough for IDs, this approach makes most use cases more efficient.

Reply:
> Good point! My primary design principle is to reduce storage overhead as much as possible. An ordinary Java string uses 2 bytes for 1 character. UTF-8, even worse, uses 1 to 4 bytes per character. Since we only allow ASCII characters, we only use 1 byte per character. That's at least 50% cheaper than other encoding schemes.
> 
> If one would like to use any character other than visible ASCII characters, they could use base64 encoding on the app side, or they could just use a byte array (once we support it). Since ASCII characters are usually enough for IDs, this approach makes most use cases more efficient.

Yes, but in UTF-8 if you use only ASCII characters in your String then every character will take 1 byte. As far as I understand the the main purpose of using the highest bit.
I.e. ASCII originally uses 7 bits for 128 characters.
UTF-8 also uses 8 bits for ASCII characters because one bit is used to understand that it’s non-ASCII character which requires a second byte or more.
In other words:
“abcd” - takes 4 bytes in ASCII encoding.
“abcd” - takes 4 bytes in UTF-8 encoding.
base64 string will take the same size if encoded in ASCII or UTF-8. 
I hardly understand the reason of using ASCII encoding over UTF-8 encoding in any case (unless you specifically want to forbid characters from non-ASCII table).
Again, I’m OK with ASCII characters. It’s just interesting that we specifically forbid UTF encoding. Most likely I am just missing some requirements regarding data keys encoding due to lack of knowledge of that part of the codebase.

Reply:
About ordinary Java String. It uses UTF-16 by default. I definitely not suggesting using UTF-16, but UTF-8 is a better alternative then ASCII encoding in my understanding almost in any case due to taking the same amount of storage for ASCII characters, but allowing to use more characters (with less optimal encoding) in case the user needs to use those characters.

Reply:
> Thanks, that's something I wasn't familiar with and very interesting.
> 
> It seems to me that the biggest problem of using UTF-8 (rather than ASCII) is we don't know the end of the string. Recall that when vertex id is stored in `column` column or `value` column, it's only part of the whole byte buffer and JanusGraph must find out the boundary by itself. When we use ASCII only, we know the first bit is always 0, thus we can mark the first bit of the last byte to be 1 (the same trick as encoding a long value). This trick does not work for UTF-8 because the first bit of a UTF-8 character can be 1.

Yeah, I now see what you mean. I assume this could be needed for index store (probably but not sure).
As for edges `column` key, to get `adjacent vertex id` from an edge, I thought it could be possible to determine key by getting bytes between `sort key` (or label id + direction) and `edge id`, because we know that `edge id` is of size 8 bytes, so the `adjacent vertex id` should be everything after `sort key` and up to `edge id`. That said, this will work to retrieve key from an edge, but I'm not sure it will work to retrieve key from other stores.
I don't have immediate solution on top of my head on how to allow using UTF-8  encoding for index store without using additional 4 bytes at the beginning to determine string length. Thus, I'm good with ASCII encoding.
I think if we implement `byte[]` later then users can use that structure to store UTF-8 strings if they really want to use non-ASCII chars for their vertex ids (which is probably a rare case).

Reply:
> I thought it could be possible to determine key by getting bytes between sort key (or label id + direction) and edge id

This may or may not work; in some cases (depending on multiplicity and constraints), the adjacent vertex id is part of `value` column followed by properties. The above trick might fail in this case.

Another perspective is: what if we allow custom edge id as the next step? Then you cannot rely on the fact that edge id is 8 bytes.

> I think if we implement byte[] later then users can use that structure to store UTF-8 strings if they really want to use non-ASCII chars for their vertex ids (which is probably a rare case).

Yep, or they can base64 encode their vertex ids first.

Reply:
> > I thought it could be possible to determine key by getting bytes between sort key (or label id + direction) and edge id
> 
> This may or may not work; in some cases (depending on multiplicity and constraints), the adjacent vertex id is part of `value` column followed by properties. The above trick might fail in this case.
> 
> Another perspective is: what if we allow custom edge id as the next step? Then you cannot rely on the fact that edge id is 8 bytes.
> 
> > I think if we implement byte[] later then users can use that structure to store UTF-8 strings if they really want to use non-ASCII chars for their vertex ids (which is probably a rare case).
> 
> Yep, or they can base64 encode their vertex ids first.

Yeah, that make sense definitely and relying on `edge id` size is a bad idea as you mentioned because we may change edge ids later on as well.
I think using ASCII only characters is a good solution here.